modules/holodeckb2b-core/src/test/java/org/holodeckb2b/security/handlers/CreateWSSHeadersTest.java

```diff
@@ -0,0 +1,270 @@
+    static final QName USERNAME_TOKEN_ELEMENT_NAME =
+            new QName(SecurityConstants.WSS_NAMESPACE_URI, "UsernameToken");
+    static final QName USERNAME_ELEMENT_NAME =
+            new QName(SecurityConstants.WSS_NAMESPACE_URI, "Username");
+    static final QName PASSWORD_ELEMENT_NAME =
+            new QName(SecurityConstants.WSS_NAMESPACE_URI, "Password");
+
+    private static String baseDir;
+
+    private static HolodeckCore core;
+
+    private CreateWSSHeaders handler;
+
+    @BeforeClass
+    public static void setUpClass() {
+        baseDir = CreateWSSHeadersTest.class
+                .getClassLoader().getResource("security").getPath();
+        core = new HolodeckCore(baseDir);
+        HolodeckB2BCoreInterface.setImplementation(core);
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        handler = new CreateWSSHeaders();
+    }
+
+    @Test
+    public void testDoProcessing () throws Exception {
+        System.out.println("[testDoProcessing]>");
+        final String mmdPath =
+                this.getClass().getClassLoader()
+                        .getResource("security/handlers/full_mmd.xml").getPath();
+        final File f = new File(mmdPath);
+        MessageMetaData mmd = null;
+        try {
+            mmd = MessageMetaData.createFromFile(f);
+        } catch (final Exception e) {
+            fail("Unable to test because MMD could not be read correctly!");
+        }
+        // Creating SOAP envelope
+        SOAPEnvelope env =
+                SOAPEnv.createEnvelope(SOAPEnv.SOAPVersion.SOAP_12);
+        // Adding header
+        SOAPHeaderBlock headerBlock = Messaging.createElement(env);
+        // Adding UserMessage from mmd
+        UserMessage.createElement(headerBlock, mmd);
+
+        MessageContext mc = new MessageContext();
+        mc.setFLOW(MessageContext.OUT_FLOW);
+        mc.setProperty(SecurityConstants.ADD_SECURITY_HEADERS, Boolean.TRUE);
+
+        // Setting signature configuration
+        SigningConfig sigConfig = new SigningConfig();
+        sigConfig.setKeystoreAlias("exampleca");
+        sigConfig.setCertificatePassword("ExampleCA");
+
+        // Setting encription configuration
+        EncryptionConfig encConfig = new EncryptionConfig();
+        encConfig.setKeystoreAlias("partyb");
+        encConfig.setCertificatePassword("ExampleB");
+
+        // Setting token configuration
+        UsernameTokenConfig tokenConfig = new UsernameTokenConfig();
+        tokenConfig.setUsername("username");
+        tokenConfig.setPassword("secret");
+
+        mc.setProperty(SecurityConstants.SIGNATURE, sigConfig);
+        // First of all we need to test RaiseSignatureCreatedEvent
+        // So, we need only signature headers by now
+        // todo Setting encription causes
+        // todo org.apache.xml.security.exceptions.XMLSecurityException
+        // todo for unknown reason. I'll check this later (T.S.)
```

Comment:
The exception can be caused because the security provider is not Bouncy Caste when the test is run. 
The default encryption algorithm is AES128-GCM which requires a different initialization which Bouncy Castle handles internally, but the default Sun provider does not.